lib/puppet/functions/lookup.rb

```diff
@@ -1,153 +1,128 @@
-#
-# For more options and details about types, see the Puppet Language Reference
-#
-# *Handling of undef*
-#
-# When no match is found for the given `name` when searching all sources, (including the `override`and
-# `default_values_hash`), then the value used is either the `default_value` or the value produced by the given block.
-# If neither is provided, then the lookup will always raise an error. Note that this only applies when there's no
-# match for the given `name`. It does not happen when a value is found and that value happens to be _undef_.
-#
-# *Validation of returned value*
-#
-# The produced value is subject to type validation using the `value_type` (if given) and an error is raised unless
-# the resulting value is of correct type.
-#
-# *Examples*
-#
-# When called with one argument; **the name**, it
-# returns the bound value with the given name after having  asserted it has the default datatype `Data`:
-#
-#     lookup('the_name')
-#
-# When called with two arguments; **the name**, and **the expected type**, it
-# returns the bound value with the given name after having asserted it has the given data
-# type ('String' in the example):
-#
-#     lookup('the_name', String)
-#
-# When called with four arguments, **the name**, the **expected type**, the **merge** strategy, and a
-# **default value**, it returns the bound value with the given name, or the default after having asserted the value
-# has the given data type:
-#
-#     lookup('the_name', String, undef, 'Fred')
-#     lookup('the_name', Array[String], 'unique', [Fred])
-#
-# Using a lambda to provide a default value by calling a function:
-#
-#     lookup('the_size', Integer[1,100]) |$name| {
-#       obtain_size_default()
-#     }
-#
-# There are two ways to make lookup return undef when no matching key was found instead of raising an error.
-# Either call it with four arguments (the `merge` argument must be present even when using the default strategy
-# to ensure that the four argument variant is used):
-#
-#      $are_you_there = lookup('peekaboo', Optional[String], undef, undef)
-#
-# or call it using an options hash:
-#
-#      $are_you_there = lookup('peekaboo', { 'default_value' => undef })
-#      $are_you_there = lookup({ 'name' => 'peekaboo', 'default_value' => undef })
-#
-# or with a block that produces an undef value:
-#
-#      $are_you_there = lookup('peekaboo', Optional[String]) |$name| { undef }
-#
-# - Since 4.0.0
+# Uses the Puppet lookup system to retrieve a value for a given key. By default,
+# this returns the first value found (and fails compilation if no values are
+# available), but you can configure it to merge multiple values into one, fail
+# gracefully, and more.
+#
+# When looking up a key, Puppet will search up to three tiers of data, in the
+# following order:
+#
+# 1. Hiera.
+# 2. The current environment's data provider.
+# 3. The indicated module's data provider, if the key is of the form
+#    `<MODULE NAME>::something`.
```

Comment:
Can we use a capped bracketed keyword (maybe <KEY>?) in place of "something" for consistency?